stm/stm.py

```diff
@@ -0,0 +1,418 @@
+import xarray as xr
+from rasterio import features
+import math
+import numpy as np
+import geopandas as gpd
+import affine
+from pathlib import Path
+from shapely.strtree import STRtree
+from shapely.geometry import Point
+from collections.abc import Iterable
+import dask.array as da
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+# Inspiration:
+#   - https://docs.xarray.dev/en/stable/internals/extending-xarray.html
+#   - https://corteva.github.io/rioxarray/html/_modules/rioxarray/raster_dataset.html
+
+
+@xr.register_dataset_accessor("stm")
+class SpaceTimeMatrix:
+    """
+    Space-Time Matrix
+    """
+
+    def __init__(self, xarray_obj):
+        self._obj = xarray_obj
+
+    def add_metadata(self, metadata):
+        """
+        Assign metadata to the STM.
+
+        Parameters
+        ----------
+        metadata : str or dict
+            input metadata
+
+        Returns
+        -------
+        xarray.Dataset
+            STM with assigned attributes.
+        """
+        self._obj = self._obj.assign_attrs(metadata)
+        return self._obj
+
+    def subset(self, method: str, **kwargs):
+        """
+        Select a subset of the STM
+
+        Parameters
+        ----------
+        method : str
+            Method of subsetting. Choose from "threshold", "density" and "polygon".
+            - threshold: select all points with a threshold criterion, e.g.
+                data_xr.stm.subset(method="threshold", var="thres", threshold='>1')
+            - density: select one point in every [dx, dy] cell, e.g.
+                data_xr.stm.subset(method='density', dx=0.1, dy=0.1)
+            - polygon: select all points inside a given polygon, e.g.
+                data_xr.stm.subset(method='polygon', polygon=path_polygon_file)
+                or
+                import geopandas as gpd
+                polygon = gpd.read_file(path_polygon_file)
+                data_xr.stm.subset(method='polygon', polygon=polygon)
+
+        Returns
+        -------
+        xarray.Dataset
+            A subset of the original STM.
+        """
+
+        match method:  # Match statements available only from python 3.10 onwards
```

Comment:
I still think too much of each subset routine's base code is within the routine. I would suggest trying to move that out, only making a call from the subset function. This would also help with future additions  of subset methods

Reply:
converted to #16